src/main/scala/com/microsoft/hyperspace/index/rules/PEFilterIndexRule.scala

```diff
@@ -0,0 +1,161 @@
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InMemoryFileIndex, LogicalRelation}
+import org.apache.spark.sql.types.StructType
+
+import com.microsoft.hyperspace.{ActiveSparkSession, Hyperspace}
+import com.microsoft.hyperspace.actions.Constants
+import com.microsoft.hyperspace.index.{IndexConstants, IndexLogEntry}
+import com.microsoft.hyperspace.index.rankers.FilterIndexRanker
+import com.microsoft.hyperspace.telemetry.HyperspaceEventLogging
+import com.microsoft.hyperspace.util.ResolverUtils
+
+/**
+ * FilterIndex rule looks for opportunities in a logical plan to replace
+ * a relation with an available covering index according to columns in
+ * filter predicate.
+ */
+object PEFilterIndexRule
+    extends Rule[LogicalPlan]
+    with Logging
+    with HyperspaceEventLogging
+    with ActiveSparkSession {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transformDown {
+      case ExtractFilterNode(originalPlan, filter, _, filterColumns) =>
+        try {
+          val candidateIndexes = findCoveringIndexes(filter, filterColumns)
+          FilterIndexRanker.rank(spark, filter, candidateIndexes) match {
+            case Some(index) => transformPlan(filter, index, originalPlan)
+            case None => originalPlan
+          }
+        } catch {
+          case e: Exception =>
+            logWarning("Non fatal exception in running filter index rule: " + e.getMessage)
+            originalPlan
+        }
+    }
+  }
+
+  def transformPlan(
+      filter: Filter,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan): LogicalPlan = {
+    // Filter useable files from original data source.
+    val fileList = getFilteredFiles(filter, index)
+
+    // Make InMemoryFileIndex from file list.
+    val provider = Hyperspace.getContext(spark).sourceProviderManager
+    val returnVal = originalPlan transformDown {
+      case l @ LogicalRelation(
+            fs @ HadoopFsRelation(oldFileIndex: InMemoryFileIndex, _, _, _, _, _),
+            _,
+            _,
+            _) =>
+        val relation = provider.getRelation(l)
+        val options = relation.partitionBasePath
+          .map { basePath =>
+            // Set "basePath" so that partitioned columns are also included in the output schema.
+            Map("basePath" -> basePath)
+          }
+          .getOrElse(Map())
+        val schema = StructType((l.schema ++ oldFileIndex.partitionSchema).distinct)
+        val newFileIndex = new InMemoryFileIndex(spark, fileList, options, Some(schema))
+        val fsOptions = fs.options ++ Map(IndexConstants.INDEX_RELATION_IDENTIFIER)
+
+        l.copy(fs.copy(location = newFileIndex, options = fsOptions)(spark))
+    }
+
+    returnVal
+  }
+
+  def getFilteredFiles(filter: Filter, index: IndexLogEntry): Seq[Path] = {
+    val condition = getIndexCompatibleCondition(filter.condition, index)
+
+    val filteredDf =
+      spark.read
+        .parquet(index.content.files.map(_.toString): _*)
```

Comment:
I thought we build an index of (value => file id list) instead of full scan of covering index.
Could you measure the perf of optimize phase with 1TB TPCH dataset?

Reply:
this would not work if we want to merge Covering and partition elimination index.

Reply:
We cannot deliver this rule without a proper performance validation as it seems expensive. Could you run with TPCH 1TB 100k chunk dataset and share the result? 

1) w/o FEFilterIndexRule - explain time & query execution time
2) w/ PEFilterindexRule - explain time & query execution time